[VPlan] Introduce scalar loop header in plan, remove VPLiveOut.

fhahn opened the pull request:
Update VPlan to include the scalar loop header. This allows retiring VPLiveOut, as the remaining live-outs can now be handled by adding operands to the wrapped phis in the scalar loop header.

Note that the current version only includes the scalar loop header, no other loop blocks and also does not wrap it in a region block.

fhahn committed:
[VPlan] Introduce scalar loop header in plan, remove VPLiveOut.

Update VPlan to include the scalar loop header. This allows retiring
VPLiveOut, as the remaining live-outs can now be handled by adding
operands to the wrapped phis in the scalar loop header.

Note that the current version only includes the scalar loop header, no
other loop blocks and also does not wrap it in a region block. This can
either be included in this PR or in follow-ups as needed.

david-arm commented:
I just noticed this patch today. I have a patch to enable vectorisation of more early exit loops that require adding live outs from the early exiting block - https://github.com/llvm/llvm-project/pull/88385. I hadn't realised that VPLiveOuts were going to be removed from the vplan.

@fhahn If possible, can you point me at the patch or code that shows how we model outside uses of loop values in the exit block? I will need to update PR #88385 to ensure we generate the correct final value in the early exit block.

david-arm commented:
Also, is there a document somewhere that highlights (even at a high level) some of the refactoring work like this that is planned in the next few months? Just because I'm also actively working in this area. Or if not, is it possible to tag me as a reviewer or subscriber for other patches in this area? Thanks!

fhahn commented:
> I just noticed this patch today. I have a patch to enable vectorisation of more early exit loops that require adding live outs from the early exiting block - #88385. I hadn't realised that VPLiveOuts were going to be removed from the vplan.
> 
> @fhahn If possible, can you point me at the patch or code that shows how we model outside uses of loop values in the exit block? I will need to update PR #88385 to ensure we generate the correct final value in the early exit block.

I think we should already have most of the pieces needed to model the exit values in the early exit the same way as for the regular exits after introducing VPIRBasicBlocks and VPIRInstructions. https://github.com/llvm/llvm-project/pull/100735 uses those for the regular exit phis.

It should be possible create a VPIRBasicBlock for the early exit block in a similar fashion and add operands to the VPIRInstructions of the phis that need updating.

A few small changes may be needed to support VPIRBBs and VPIRInst phis with multiple predecessors, but hopefully something like 3cfe25bf43feaf94d319c7f804e5c73c45f96057 should be enough.

fhahn commented:
>  Or if not, is it possible to tag me as a reviewer or subscriber for other patches in this area? Thanks!

Probably the best way would be to join the subscriber team for the vectorizer label: https://github.com/orgs/llvm/teams/pr-subscribers-vectorizers

david-arm commented:
> > I just noticed this patch today. I have a patch to enable vectorisation of more early exit loops that require adding live outs from the early exiting block - #88385. I hadn't realised that VPLiveOuts were going to be removed from the vplan.
> > @fhahn If possible, can you point me at the patch or code that shows how we model outside uses of loop values in the exit block? I will need to update PR #88385 to ensure we generate the correct final value in the early exit block.
> 
> I think we should already have most of the pieces needed to model the exit values in the early exit the same way as for the regular exits after introducing VPIRBasicBlocks and VPIRInstructions. #100735 uses those for the regular exit phis.
> 
> It should be possible create a VPIRBasicBlock for the early exit block in a similar fashion and add operands to the VPIRInstructions of the phis that need updating.
> 
> A few small changes may be needed to support VPIRBBs and VPIRInst phis with multiple predecessors, but hopefully something like [3cfe25b](https://github.com/llvm/llvm-project/commit/3cfe25bf43feaf94d319c7f804e5c73c45f96057) should be enough.

OK thank you for the info - I'll take a look!

david-arm commented:
> > Or if not, is it possible to tag me as a reviewer or subscriber for other patches in this area? Thanks!
> 
> Probably the best way would be to join the subscriber team for the vectorizer label: https://github.com/orgs/llvm/teams/pr-subscribers-vectorizers

Yeah, I remember when we first moved to github I added myself to that list and I didn't seem to get any notifications of vectoriser patches, despite it being pretty obvious that patches were being created. So I took the alternative approach of saving a link to a github search query that shows all open vectoriser patches. Maybe it has started working now?

fhahn commented:
> > > I just noticed this patch today. I have a patch to enable vectorisation of more early exit loops that require adding live outs from the early exiting block - #88385. I hadn't realised that VPLiveOuts were going to be removed from the vplan.
> > > @fhahn If possible, can you point me at the patch or code that shows how we model outside uses of loop values in the exit block? I will need to update PR #88385 to ensure we generate the correct final value in the early exit block.
> > 
> > 
> > I think we should already have most of the pieces needed to model the exit values in the early exit the same way as for the regular exits after introducing VPIRBasicBlocks and VPIRInstructions. #100735 uses those for the regular exit phis.
> > It should be possible create a VPIRBasicBlock for the early exit block in a similar fashion and add operands to the VPIRInstructions of the phis that need updating.
> > A few small changes may be needed to support VPIRBBs and VPIRInst phis with multiple predecessors, but hopefully something like [3cfe25b](https://github.com/llvm/llvm-project/commit/3cfe25bf43feaf94d319c7f804e5c73c45f96057) should be enough.
> 
> OK thank you for the info - I'll take a look!

https://github.com/llvm/llvm-project/pull/109193/files#diff-c33c5be02bdbedc5499230934c8fc873429b771425f21c3f594acb6fae6339d7 should show how the VPlans would look like

david-arm reviewed (dismissed):
Hi @fhahn, whilst the PR looks fine in isolation I'd rather it not land until we've had a chance to discuss its impact on PR #88385 if that's ok? Thanks!

fhahn commented:
Sure. Curious if you hit any limitations with VPIRInstruction that prevent it from replacing VPLiveOut?

david-arm commented:
> Sure. Curious if you hit any limitations with VPIRInstruction that prevent it from replacing VPLiveOut?

Yeah, so in my comment earlier last week I highlighted the problem of ordering constraints. If the same exit block is used by both the latch and the early exit, then the phi in the exit block should have an incoming value for both of the exiting blocks. If I add operands to the VPIRInstruction wrapper - one for each exiting block - then in the VPIRInstruction::execute code using a patch you sketched out in https://github.com/llvm/llvm-project/commit/3cfe25bf43feaf94d319c7f804e5c73c45f96057 we introduce a requirement that the order of the extra operands *must* match the order of the predecessors. `replaceVPBBWithIRVPBB` breaks this by reordering them, hence I put up a patch yesterday to preempt this - https://github.com/llvm/llvm-project/pull/111514. Perhaps it's fine to introduce this ordering constraint, but I wonder if this makes the code fragile and prone to breaking every time we do some CFG optimisation? Secondly, the phi may or may not reference an induction variable in the loop, which is handled separately to VPIRInstruction::execute. However, you still need to add some sort of dummy operand because the number of operands *must* match the number of predecessors. I have hacked this downstream by adding a static 'null' VPValue marker that we can skip over the operand in VPIRInstruction::execute.

Another approach would be to introduce a new class derived from VPIRInstruction, i.e. VPIRPhiInstruction that maintains a mapping between the extra operand and the incoming block. I haven't tried this second approach - this is potentially more flexible, but it is moving a bit further away from having a simple IR wrapper and is moving back towards something like VPLiveOut.

fhahn commented:
> > Sure. Curious if you hit any limitations with VPIRInstruction that prevent it from replacing VPLiveOut?
> 
> Yeah, so in my comment earlier last week I highlighted the problem of ordering constraints. If the same exit block is used by both the latch and the early exit, then the phi in the exit block should have an incoming value for both of the exiting blocks. If I add operands to the VPIRInstruction wrapper - one for each exiting block - then in the VPIRInstruction::execute code using a patch you sketched out in [3cfe25b](https://github.com/llvm/llvm-project/commit/3cfe25bf43feaf94d319c7f804e5c73c45f96057) we introduce a requirement that the order of the extra operands _must_ match the order of the predecessors. `replaceVPBBWithIRVPBB` breaks this by reordering them, hence I put up a patch yesterday to preempt this - #111514. Perhaps it's fine to introduce this ordering constraint, but I wonder if this makes the code fragile and prone to breaking every time we do some CFG optimisation? 

The order of successors/predecessors is already meaningful in VPlan, it is e.g. also used determine to true/false successors for branches. So far, `replaceVPBBWithIRVPBB` wasn't an issue, because of the middle vpbb only have a single predecessor. Ensuring that more generally via  #111514 seems like the right way to go, thanks for putting up the patch.

> Secondly, the phi may or may not reference an induction variable in the loop, which is handled separately to VPIRInstruction::execute. However, you still need to add some sort of dummy operand because the number of operands _must_ match the number of predecessors. I have hacked this downstream by adding a static 'null' VPValue marker that we can skip over the operand in VPIRInstruction::execute.
> 

Yes, fixing induction users in the exit block is the last remaining exit users handling not managed by VPlan, but this should also be migrated soon. I shared a stack of patches that move computing the end value to VPlan (https://github.com/llvm/llvm-project/pull/112145) and updating the exit users of IVs, retiring fixupIVUsers (https://github.com/llvm/llvm-project/pull/112147). They both depend on this patch and https://github.com/llvm/llvm-project/pull/110577, but should give a good idea on what is needed to complete the transition of modeling all live-outs directly in VPlan.

fhahn commented:
Just updated the patch, including
* updating all required test
* updated VPlan to hold a pointer to the (unchangeable) scalar header VPIRBasicBlock and added helpers to retrieve itself and the scalar preheader using it, addressing another TODO

ping :)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1260,10 +1261,12 @@ static void remapOperands(VPBlockBase *Entry, VPBlockBase *NewEntry,
   // Create VPlan, clone live-ins and remap operands in the cloned blocks.
-  auto *NewPlan = new VPlan(NewPreheader, cast<VPBasicBlock>(NewEntry));
+  auto *NewPlan =
+      new VPlan(NewPreheader, cast<VPBasicBlock>(NewEntry), NewScalarHeader);
```
nit (unrelated): should NewEntry be returned as a VPBB rather than a VPBlockBase?

fhahn replied:
Will do separately, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -675,48 +675,6 @@ class VPBlockBase {
-/// determined by where the VPValue is defined: if it is defined by a recipe
-/// outside a region, its parent block is used, otherwise the middle block is
-/// used.
-class VPLiveOut : public VPUser {
```
Note that with VPLiveOut replaced by VPIRInstructions, all VPUsers are recipes.

fhahn replied:
Yep, removed `LiveOut` VPUserID, but retained the enum for now (may be helpful for marking VPBlocks/Regions as users (e.g. of VF, TC). But can also drop it all and add back if/when needed.

ayalz replied:
Dropping seems cleaner to me.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3642,22 +3598,26 @@ class VPlan {
   /// and \p Entry to the plan. At the moment, \p Preheader and \p Entry need to
```
Augment documentation with additional \p.

fhahn replied:
Done, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3642,22 +3598,26 @@ class VPlan {
   }
 
   /// Construct a VPlan with original preheader \p Preheader and \p Entry to
   /// the plan. At the moment, \p Preheader and \p Entry need to be
```
Augment documentation with additional \p.

fhahn replied:
Done, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -869,7 +840,10 @@ void VPIRInstruction::execute(VPTransformState &State) {
     Value *V = State.get(ExitValue, VPLane(Lane));
     auto *Phi = cast<PHINode>(&I);
-    Phi->addIncoming(V, PredBB);
+    if (Phi->getBasicBlockIndex(PredBB) == -1)
```
Worth commenting about the two types of VPIRBB's - those that add a new ("extra") incoming and those that update an existing one?

fhahn replied:
Updated to add a comment, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -379,7 +379,7 @@ static bool mergeBlocksIntoPredecessors(VPlan &Plan) {
     // now.
     // TODO: Remove restriction once more of the skeleton is modeled in VPlan.
-    if (VPBB->getNumSuccessors() == 0 && !VPBB->getParent())
+    if (!VPBB->getParent())
```
The condition for zero successors is dropped because it now coincides with (scalar header and exit) having no parental region? Above comment about refraining from folding exit block "for now" still intact?

fhahn replied:
Yep, updated to skip this for any block in the skeleton.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanUnroll.cpp:
```diff
@@ -264,6 +264,13 @@ void UnrollState::unrollRecipeByUF(VPRecipeBase &R) {
 
   if (auto *VPI = dyn_cast<VPInstruction>(&R)) {
+    VPValue *Op0, *Op1;
+    if (match(VPI, m_VPInstruction<VPInstruction::ResumePhi>(m_VPValue(Op0),
```
How/Is this change related to this patch, ResumePhi's for FORs are used by other than part zero?

fhahn replied:
This is related to removing VPLiveOut, VPIRInstructions aren't marked as using the first lane part/lane (technically they would use the last part/lane?)

ayalz replied:
VPInstruction::ResumePhi `isSingleScalar()`, so should probably qualify as only first part used below?

ayalz reviewed (commented):
Very nice cleanup!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -663,14 +670,16 @@ void VPBasicBlock::print(raw_ostream &O, const Twine &Indent,
 
 // Clone the CFG for all nodes reachable from \p Entry, this includes cloning
 // the blocks and their recipes. Operands of cloned recipes will NOT be updated.
 // Remapping of operands must be done separately. Returns a pair with the new
```
Removed related changes, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1054,6 +1064,8 @@ void VPlan::execute(VPTransformState *State) {
   BrInst->insertBefore(MiddleBB->getTerminator());
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1054,6 +1064,8 @@ void VPlan::execute(VPTransformState *State) {
   MiddleBB->getTerminator()->eraseFromParent();
   State->CFG.DTU.applyUpdates({{DominatorTree::Delete, MiddleBB, ScalarPh}});
+  State->CFG.DTU.applyUpdates(
+      {{DominatorTree::Delete, ScalarPh, ScalarPh->getSingleSuccessor()}});
```
Yep

fhahn replied:
Added, thanks!

fhahn commented on llvm/unittests/Transforms/Vectorize/VPDomTreeTest.cpp:
```diff
@@ -41,7 +41,11 @@ TEST(VPDominatorTreeTest, DominanceNoRegionsTest) {
 
   auto TC = std::make_unique<VPValue>();
-  VPlan Plan(VPPH, &*TC, VPBB0);
+  LLVMContext C;
+  auto ScalarHeader = std::make_unique<BasicBlock *>(BasicBlock::Create(C, ""));
+  VPIRBasicBlock *ScalarHeaderVPBB = new VPIRBasicBlock(*ScalarHeader);
```
I'm not entirely sure how merging them would help in this particular case? We *could* make the scalar header optional, but it seems good to ensure it is always set.

ayalz replied:
Ah, right, ok.
Perhaps worth connecting ScalarHeaderVPBB to the other VPBB's and have it participate in dominance checks.

fhahn committed:
!fixup address latest comments, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3746,6 +3708,14 @@ class VPlan {
   const VPBasicBlock *getEntry() const { return Entry; }
 
+  /// Return the VPIRBasicBlock wrapping the header of the scalar loop.
+  VPIRBasicBlock *getScalarHeader() { return ScalarHeader; }
```
const?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3746,6 +3708,14 @@ class VPlan {
+  VPIRBasicBlock *getScalarHeader() { return ScalarHeader; }
+
+  /// Return the VPBasicBlock for the preheader of the scalar loop.
+  VPBasicBlock *getScalarPreheader() {
```
const?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -376,10 +376,10 @@ static bool mergeBlocksIntoPredecessors(VPlan &Plan) {
            vp_depth_first_deep(Plan.getEntry()))) {
-    // Don't fold the exit block of the Plan into its single predecessor for
-    // now.
+    // Don't fold the blocks in the skeleton of the Plan into its single
```
```suggestion
    // Don't fold the blocks in the skeleton of the Plan into their single
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanValue.h:
```diff
@@ -202,7 +202,6 @@ class VPUser {
   /// Subclass identifier (for isa/dyn_cast).
   enum class VPUserID {
     Recipe,
-    LiveOut,
   };
```
Perhaps better clean this up for now, to avoid false pretense it is in use, and reintroduce if/when desired (for VF/TC).

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-liveout

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-liveout

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8978,15 +8962,25 @@ static void addLiveOutsForFirstOrderRecurrences(
     auto *FORPhi = cast<PHINode>(FOR->getUnderlyingInstr());
-    Plan.addLiveOut(FORPhi, ResumePhiRecipe);
+    VPIRInstruction *IRI = nullptr;
+    for (VPRecipeBase &R : *Plan.getScalarHeader()) {
```
Changed the approach now that ResumePHI are needed for both FORs and reductions: 

added `addScalarResumePhis` which creates the ResumePhi recipes for both cases (and eventually inductions as well). They are added as users of the backedge value, and later will be updated by replacing users of the backedge value (it works for reductions without any changes, for recurrences we replace uses in the scalar preheader)

WDYT?

fhahn committed:
!fixup address latest comments, thanks!

ayalz reviewed (approved):
This looks fine to me, adding various thoughts and tentative suggestions.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-remove-liveout

fhahn committed:
!fixup address comments, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -455,8 +455,9 @@ void VPIRBasicBlock::execute(VPTransformState *State) {
-  if (getSingleSuccessor()) {
-    assert(isa<UnreachableInst>(IRBB->getTerminator()));
+  // Create a branch instruction to terminate IRBB if one was not created yet
+  // and is needed.
```
Added thanks

fhahn committed:
!drop unused variable

fhahn committed:
!fixup try to fix windows build failure

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9166,11 +9174,11 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
   SetVector<VPIRInstruction *> ExitUsersToFix = collectUsersInExitBlock(
       OrigLoop, RecipeBuilder, *Plan, Legal->getInductionVars());
   addLiveOutsForFirstOrderRecurrences(*Plan, ExitUsersToFix);
   addUsersInExitBlock(*Plan, ExitUsersToFix);
```
Adjusted as suggested, will revisit and further unify as follow-up (at the latest when inductions are also modeled in VPlan)

fhahn committed:
!fixup document reason for disconnecting scalar ph and header in DTU.

kstoimenov commented:
@fhahn you might have introduced a memory leak in this patch detected by Sanitizer bots: https://lab.llvm.org/buildbot/#/builders/169/builds/4844/steps/10/logs/stdio

Please take a look.

fhahn commented:
Should be fixed in 4bc8bde0bc4b0dbace5545de5792be2138e9cf64, thanks!

kstoimenov commented:
Looks like there is still one more leak left after your fix: https://lab.llvm.org/buildbot/#/builders/169/builds/4871/steps/10/logs/stdio

kstoimenov commented:
FYI, @vitalybuka pushed this fix: 80b9f074
